pkg/controller/quota/quota.go

```diff
@@ -0,0 +1,168 @@
+		Spec: adminv1.QuotaRequestInstanceSpec{
+			Resources: adminv1.Resources{
+				Containers: len(app.Containers),
+				Jobs:       len(app.Jobs),
+				Volumes:    len(app.Volumes),
+				Secrets:    len(app.Secrets),
+				Images:     len(app.Images),
+			},
+		},
+	}
+
+	status := condition.Setter(appInstance, resp, apiv1.AppInstanceConditionQuotaAllocated)
+
+	// Add the more complex values to the quota request
+	addCompute(app.Containers, appInstance, quotaRequest)
+	addCompute(app.Jobs, appInstance, quotaRequest)
+	if err := addStorage(appInstance, quotaRequest); err != nil {
+		status.Error(err)
+		return err
+	}
+
+	resp.Objects(quotaRequest)
+	return nil
+}
+
+// addCompute adds the compute resources of the containers passed to the quota request.
+func addCompute(containers map[string]apiv1.Container, appInstance *apiv1.AppInstance, quotaRequest *adminv1.QuotaRequestInstance) {
+	// For each workload, add their memory/cpu requests to the quota request
+	for name, container := range containers {
+		var requirements corev1.ResourceRequirements
+		if specific, ok := appInstance.Status.Scheduling[name]; ok {
+			requirements = specific.Requirements
+		} else if all, ok := appInstance.Status.Scheduling[""]; ok {
+			requirements = all.Requirements
+		}
+
+		quotaRequest.Spec.Resources.CPU.Add(requirements.Requests["cpu"])
+		quotaRequest.Spec.Resources.Memory.Add(requirements.Requests["memory"])
+
+		// Recurse over any sidecars. Since sidecars can't have sidecars, this is safe.
+		addCompute(container.Sidecars, appInstance, quotaRequest)
+	}
+}
+
+// addStorage adds the storage resources of the volumes passed to the quota request.
+func addStorage(appInstance *apiv1.AppInstance, quotaRequest *adminv1.QuotaRequestInstance) error {
+	app := appInstance.Status.AppSpec
+
+	// Add the volume storage needed to the quota request. We only parse net new volumes, not
+	// existing ones that are then bound client-side.
+	for name, volume := range app.Volumes {
+		size := volume.Size
+		if bound, boundSize := boundVolumeSize(name, appInstance.Spec.Volumes); bound {
+			size = boundSize
+		}
+
+		// No need to proceed if the size is empty
+		if size == "" || size == "0" {
+			continue
+		}
+
+		parsedSize, err := resource.ParseQuantity(string(size))
+		if err != nil {
+			return err
+		}
+		quotaRequest.Spec.Resources.VolumeStorage.Add(parsedSize)
+	}
+	return nil
+}
+
+// TODO - Need function here for determining if a secret will be bound as well
+
+func boundVolumeSize(name string, bindings []apiv1.VolumeBinding) (bool, apiv1.Quantity) {
+	for _, binding := range bindings {
+		if binding.Target == name && binding.Volume != "" {
+			return true, binding.Size
+		}
+	}
+	return false, "0"
+}
+
+func isEnforced(req router.Request, namespace string) (bool, error) {
+	project := v1.Project{}
+	if err := req.Client.Get(req.Ctx, router.Key("", namespace), &project); err != nil {
+		if errors.IsNotFound(err) {
```

Comment:
Why do we need to handle a not found error differently?

Reply:
Ah, right. I wrote this awhile ago so the current logic may get around this but this made it so that the quota request wouldn't error out if the project wasn't created yet. Basically the status would be updated with the equivalent of `project not found` for quota allocation which was a confusing UX since it happened consistently enough.

Reply:
You're saying that you were able to create an app or quota in a project that didn't exist?

Reply:
I just tested this locally and you are right, there is not a need to handle a not-found error differently here. I'll go ahead and amend this change.

Reply:
Yeah, if project is guaranteed to exist, we should not assert not found error, and just let it fail if error happens.

Reply:
I believe this not existing was leading to quite weird number of test failures. After revert it back to normal everything is passing again. Thoughts? Looking at the controller logs I saw a number of failures to sync Projects.

CC: @StrongMonkey @thedadams

Reply:
The issue here seems to be that you are looking for the project at all. Since this is in the controller, the controller shouldn't be looking for `api.acorn.io` objects. Instead, it should be looking at the underlying/internal objects. In this case, the namespace.

Reply:
Gotcha, I updated this to use a Namespace instead. Let's see if the tests get through properly.

Reply:
Yep! That seems to have fixed it.